n3fit/src/n3fit/model_gen.py

```diff
@@ -728,50 +727,48 @@ def generate_nn(
         # come from the runcard
         nodes_list[-1] = 1
         basis_size = last_layer_nodes
-        custom_args['basis_size'] = basis_size
 
-        def initializer_generator(seed, i_layer):
-            seed += i_layer * basis_size
-            initializers = [
-                MetaLayer.select_initializer(initializer_name, seed=seed + b)
-                for b in range(basis_size)
-            ]
-            return initializers
+        def layer_generator(i_layer, nodes_out, activation):
+            """Generate the ``i_layer``-th layer for all replicas."""
+            layers = []
+            for replica_seed in replica_seeds:
+                seed = replica_seed + i_layer * basis_size
+                initializers = [
+                    MetaLayer.select_initializer(initializer_name, seed=seed + b)
+                    for b in range(basis_size)
+                ]
+                layer = base_layer_selector(
+                    layer_type,
+                    kernel_initializer=initializers,
+                    units=nodes_out,
+                    activation=activation,
+                    input_shape=(nodes_in,),
+                    basis_size=basis_size,
+                )
+                layers.append(layer)
+
+            return layers
 
     else:  # "dense"
         reg = regularizer_selector(regularizer, **regularizer_args)
-        custom_args['regularizer'] = reg
 
-        def initializer_generator(seed, i_layer):
-            seed += i_layer
-            return MetaLayer.select_initializer(initializer_name, seed=seed)
-
-    # First create all the layers...
-    # list_of_pdf_layers[d][r] is the layer at depth d for replica r
-    list_of_pdf_layers = []
-    for i_layer, (nodes_out, activation) in enumerate(zip(nodes_list, activations)):
-        if layer_type == "dense":
-            layers = base_layer_selector(
+        def layer_generator(i_layer, nodes_out, activation):
+            """Generate the ``i_layer``-th layer for all replicas."""
```

Comment:
I guess just to be a pain, but since this now really only works for the MultiDense, please do `elif "dense"` so that it is obvious that this it can only enter here in the multidense case.

And maybe in the docstring of this internal function or as a comment, add something like "Generation of layer with MetaLayer.MultiDense".

This is just to avoid a future in which someone adds a new layer type and thinks this should work out of the box. The fact that the initializer don't get the seeds directly but that they go to the layer is confusing if you don't know what layer is being used here.

Reply:
But `if A: ... elif B: ...` will generate complaints from eg pylint right? That's why I just had the `else: # dense`. I can add an else that throws an exception, but I thought the policy was to have these kinds of checks all at the start through reportengine?

Reply:
Yes, the checks should be done at the beginning at it is indeed done.

But that doesn't mean you cannot throw an exception if that helps clarify the code. Here in particular someone adding a new layer type might remove the check from `checks.py` and not even read this part of the code thinking that it would work. Adding a `raise ValueError(f"{layer_type=} not recognized during model generation")` is a perfectly valid option.

Reply:
adding, just to clarify where I'm coming from, most of this comments on clarifications are because I'm thinking of the situation in which a student arrives in three years fresh from their undergraduate, with no knowledge of n3fit or probably even python and is asked to check what would happen if one uses a fancy new network architecture and check what happens.

This is a situation I obviously didn't have in mind when I first wrote n3fit (I guess I was the one ho was asked to do something and my n3fit was validphys :P) but now I'd rather have redundant explanations, checks, etc when possible rather than following a more pure approach.

Reply:
Ok, done!

Reply:
Yes it makes sense, no problem!